katdal/test/test_chunkstore_s3.py

```diff
@@ -0,0 +1,103 @@
+from nose.tools import assert_raises
+import requests
+
+import katdal.chunkstore_s3
+from katdal.chunkstore_s3 import S3ChunkStore
+
+
+class TestS3ChunkStore(object):
+    def setup(self):
+        """Start Fake S3 service running on temp dir, and ChunkStore on that."""
+        self.tempdir = tempfile.mkdtemp()
+        host = 'localhost'
+        # Choose a random port and hope it's available
+        port = np.random.randint(10000, 60000)
+        url = "http://%s:%d" % (host, port)
+        try:
+            self.fakes3 = subprocess.Popen(['fakes3', 'server', '-H', host,
+                                            '-p', str(port), '-r', self.tempdir])
+        except OSError:
+            self.fakes3 = None
+            self.teardown()
+            raise SkipTest('Could not start fakes3 server (is it installed?)')
+        start = time.time()
+        # Give up after waiting a few seconds for Fake S3
+        while time.time() - start <= 10:
+            try:
+                response = requests.get(url, timeout=0.5)
+            except requests.exceptions.ConnectionError:
+                time.sleep(0.3)
+            else:
+                # In case there is already another service on this port (even S3)
+                if response.reason == 'OK':
+                    break
+                else:
+                    self.teardown()
+                    raise SkipTest('Unexpected response from fakes3 server')
+        else:
+            self.teardown()
+            raise SkipTest('Could not connect to fakes3 server')
+        # Now start up store
+        self.x = np.arange(10)
+        self.y = np.arange(24.).reshape(4, 3, 2)
+        try:
+            self.store = S3ChunkStore(url)
+        except ImportError:
+            raise SkipTest('S3 botocore dependency not installed')
+
+    def teardown(self):
+        if self.fakes3:
+            self.fakes3.terminate()
+            self.fakes3.wait()
+        shutil.rmtree(self.tempdir)
+
+    def array_name(self, path):
+        bucket = 'katdal-unittest'
+        return self.store.join(bucket, path)
+
+    def test_store(self):
+        # Pretend that botocore is not installed
+        real_botocore = katdal.chunkstore_s3.botocore
+        katdal.chunkstore_s3.botocore = None
+        assert_raises(ImportError, S3ChunkStore, 'blah')
+        katdal.chunkstore_s3.botocore = real_botocore
+
+    def test_put_and_get(self):
+        s = (slice(3, 5),)
+        desired = self.x[s]
+        name = self.array_name('x')
+        self.store.put(name, s, desired)
+        actual = self.store.get(name, s, desired.dtype)
+        assert_array_equal(actual, desired, "Error storing x[%s]" % (s,))
+        s = (slice(1, 4), slice(1, 3), slice(1, 2))
+        desired = self.y[s]
+        name = self.array_name('y')
+        self.store.put(name, s, desired)
+        actual = self.store.get(name, s, desired.dtype)
+        assert_array_equal(actual, desired, "Error storing y[%s]" % (s,))
```

Comment:
Some other tests you should do:
- chunk doesn't exist
- S3 object has more bytes than expected
- S3 object has less bytes than expected
- `put` with Fortran order data

Reply:
The first option needs a proper exception first.

I haven't even thought of what should happen in the case of the last option...

Reply:
Done. Added tests for all chunk stores for points 1, 2 and 3.

The 4th point seems to be reasonably benign. I tried the following:

```Python
x = np.arange(24).reshape(4, 6)
x.flags.c_contiguous  # True
x.flags.f_contiguous  # False
x.T.flags.c_contiguous  # False
x.T.flags.f_contiguous  # True
s = x.T.tobytes()
y = np.ndarray((6, 4), np.int, s)
y.flags.c_contiguous   # True
y.flags.f_contiguous   # False
np.array_equal(x.T, y)   # True
```
So the array reconstructs correctly, but with the ordering fixed to 'C'. I don't store the order yet (cottoncandy does, for one), so not sure what more to do.